README.md
[![Build Status](https://travis-ci.org/kablamo/p5-shell-carapace.svg?branch=master)](https://travis-ci.org/kablamo/p5-shell-carapace) [![Coverage Status](https://img.shields.io/coveralls/kablamo/p5-shell-carapace/master.svg)](https://coveralls.io/r/kablamo/p5-shell-carapace?branch=master)
# NAME

Shell::Carapace - Simple realtime output for ssh and shell commands

# SYNOPSIS

    use Shell::Carapace::Local;

    my $callback = sub {         # require.  handles cmd output, errors, etc
        my ($category, $message) = @_;
        print "  $message\n"        if $category =~ /output/ && $message;
        print "Running $message\n"  if $category eq 'command';
        print "ERROR: cmd failed\n" if $category eq 'error';
    };

    my $shell = Shell::Carapace->shell(callback => $callback);
    $shell->run(@cmd); # throws an exception if @cmd fails

    my $ssh  = Shell::Carapace->ssh(
        callback    => $callback,    # required
        host        => $hostname,    # required
        ssh_options => $ssh_options, # a hash for Net::OpenSSH
    );
    $ssh->run(@cmd); # throws an exception if @cmd fails

# DESCRIPTION

Ever run a script that takes 30 minutes to run and have to wait
30 minutes to see the output?  This module solve that problem.

Shell::Carapace is a small wrapper around IPC::Open3::Simple and Net::OpenSSH.
It provides a callback so you can easily log or process cmd output in realtime.  

# METHODS

## shell(%options)

Creates and returns a Shell::Carapace::Shell object.  All parameters are
optional except 'callback'.  The following parameters are accepted:

    callback    : Required.  A coderef which is executed in realtime as output

## ssh(%options)

Creates and returns a Shell::Carapace::SSH object.  All parameters are optional
except 'callback'.  The following parameters are accepted:

    callback    : Required.  A coderef which is executed in realtime as output
                  is emitted from the command.
    host        : Required.  A string like 'localhost' or 'user@hostname' which
                  is passed to Net::OpenSSH.  Net::OpenSSH defaults the username
                  to the current user.
    ssh_options : A hash which is passed to Net::OpenSSH.

## $shell->run(@cmd)

Execute the command locally via IPC::Open3::Simple.  Calls the callback in
realtime for each line of output emitted from the command.

## $ssh->run(@cmd)

Execute the command on a remote host via Net::OpenSSH.  Calls the callback in
realtime for each line of output emitted from the command.

# ABOUT THE NAME

Carapace: n. A protective, shell-like covering likened to that of a turtle or crustacean

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Eric Johnson <eric.git@iijo.org>


Changes
Revision history for Shell-Carapace

{{$NEXT}}

0.22 2015-12-21T21:54:00Z
    - Another attempt to deal with open2pty() failures correctly.

0.21 2015-12-21T20:55:22Z
    - Retry if open2pty() fails and then die if it fails again.

0.20 2015-12-21T19:15:26Z
    - Last release broke handling of ssh_options attr.  This fixes it.

0.19 2015-12-21T16:16:21Z
    - Don't connect to remote host lazily -- connect on instantiation.

0.18 2015-12-15T13:46:06Z
    - Added Test::Fatal to the cpanfile
    - Made Minilla a 'develop' dependency in the cpanfile

0.17 2015-12-15T13:39:24Z
    - Die if ssh connection can't be established.

0.16 2015-11-15T23:48:14Z
    - Pass command to callback on error

0.15 2015-11-15T02:27:26Z
    - Skip t/ssh.t

0.14 2015-11-15T02:22:03Z
    - Now passing a hostname to the callback

0.13 2015-09-07T03:05:55Z
    - Major rewrite.  Not backwards compatible at all.  Splits module into two
      parts -- Shell::Carapace::Shell and Shell::Carapace::SSH.

0.12 2015-09-05T20:20:50Z
    - Major rewrite.  Not backwards compatible at all.  Introduces callback
      concept.

0.11 2015-03-24T23:45:49Z
    - Die noisely

0.10 2015-02-24T22:02:17Z
    - Made print_cmd attribute read/write.

0.09 2015-02-18T20:30:00Z
    - Tee output to the log file.  This time the exit code is preserved.

0.08 2015-02-18T15:42:16Z
    - Revert most of 0.07.  Piping to tee means we can't know what the exit
      code of the command is.  Added a test so this case isn't missed in the
      future.

0.07 2015-02-18T00:56:10Z
    - Pipe command to tee by default so logfile is written to in realtime.

0.06 2015-02-12T15:28:05Z
    - Die earlier if logfile has bad permissions

0.05 2015-02-12T00:09:13Z
    - Don't use 'coerce => 1' so Shell::Carapace is compatible with 
      Moo <1.006000

0.04 2015-02-09T18:54:46Z
    - Allow logfile attr to be undef

0.03 2015-02-09T14:25:24Z
    - Made logging to a file optional

0.02 2015-02-09T11:06:44Z
    - Minor updates to pod

0.01 2015-02-08T23:37:48Z
    - Original version



lib/Shell/Carapace/SSH.pm
package Shell::Carapace::SSH;
use Moo;

use String::ShellQuote;
use Carp;

has callback         => (is => 'rw', required => 1);
has host             => (is => 'rw', required => 1);
has ssh_options      => (is => 'rw', default => sub { {} });
has ssh              => (is => 'rw', lazy => 1, builder => 1);
has connect_attempts => (is => 'rw', default => sub { 3 });

sub _build_ssh {
    my $self = shift;
    require Net::OpenSSH;
    my %ssh_options = $self->ssh_options ? %{ $self->ssh_options } : ();
    my $ssh = Net::OpenSSH->new($self->host, %ssh_options);
    die $ssh->error if $ssh->error;
    return $ssh;
}

# force ssh builder to run so the connection occurs during object instantiation
sub BUILD { shift->ssh }

sub run {
    my ($self, @cmd) = @_;

    $self->callback->('command', $self->_stringify(@cmd), $self->host);

    my ($pty, $pid) = $self->ssh->open2pty(@cmd);
    if (!$pty || $self->ssh->error) {
        $self->clear_ssh; # force new connection to the server
        ($pty, $pid) = $self->ssh->open2pty(@cmd);
        die $self->ssh->error if $self->ssh->error;
    }

    while (my $line = <$pty>) {
      $line =~ s/([\r\n])$//g;
      $self->callback->('remote-output', $line, $self->host);
    }   

    waitpid($pid, 0);

    if ($? != 0) {
        $self->callback->("error", $self->_stringify(@cmd), $self->host);
        croak "cmd failed";
    }
}

sub _stringify {
    my ($self, @cmd) = @_;
    return $cmd[0] if @cmd == 1;
    return join(" ", shell_quote @cmd);
}

1;


lib/Shell/Carapace/Shell.pm
package Shell::Carapace::Shell;
use Moo;

use IPC::Open3::Simple;
use String::ShellQuote;
use Carp;

has callback    => (is => 'rw', required => 1);
has ipc         => (is => 'rw', lazy => 1, builder => 1);

sub _build_ipc {
    my $self = shift;
    require IPC::Open3::Simple;
    return  IPC::Open3::Simple->new(
        out => sub { $self->callback->('local-output', $_[0], 'localhost') },
        err => sub { $self->callback->('local-output', $_[0], 'localhost') },
    ); 
}

sub run {
    my ($self, @cmd) = @_;

    $self->callback->('command', $self->_stringify(@cmd), 'localhost');

    $self->ipc->run(@cmd);

    if ($? != 0) {
        $self->callback->("error", $self->_stringify(@cmd), 'localhost');
        croak "cmd failed";
    }
};

sub _stringify {
    my ($self, @cmd) = @_;
    return $cmd[0] if @cmd == 1;
    return join(" ", shell_quote @cmd);
}

1;


lib/Shell/Carapace.pm
package Shell::Carapace;
use Moo;

our $VERSION = "0.21";

=head1 NAME

Shell::Carapace - Simple realtime output for ssh and shell commands

=head1 SYNOPSIS

    use Shell::Carapace::Local;

    # A callback is required.  It can be used to log commands, output, errors
    my $shell_callback = sub {
        my ($category, $message, $host) = @_;
        print "  $host $message\n"  if $category =~ /output/ && $message;
        print "Running $message\n"  if $category eq 'command';
        print "ERROR: cmd failed\n" if $category eq 'error';
    };

    my $shell = Shell::Carapace->shell(callback => $callback);
    $shell->run(@cmd); # throws an exception if @cmd fails

    my $ssh  = Shell::Carapace->ssh(
        callback    => $callback,    # required
        host        => $hostname,    # required
        ssh_options => $ssh_options, # a hash for Net::OpenSSH
    );
    $ssh->run(@cmd); # throws an exception if @cmd fails

=head1 DESCRIPTION

Ever run a script that takes 30 minutes to run and have to wait
30 minutes to see the output?  This module solve that problem.

Shell::Carapace is a small wrapper around IPC::Open3::Simple and Net::OpenSSH.
It provides a callback so you can easily log or process cmd output in realtime.  

=head1 METHODS

=head2 shell(%options)

Creates and returns a Shell::Carapace::Shell object.  All parameters are
optional except 'callback'.  The following parameters are accepted:

   callback    : Required.  A coderef which is executed in realtime as output

=head2 ssh(%options)

Creates and returns a Shell::Carapace::SSH object.  All parameters are optional
except 'callback'.  The following parameters are accepted:

   callback    : Required.  A coderef which is executed in realtime as output
                 is emitted from the command.
   host        : Required.  A string like 'localhost' or 'user@hostname' which
                 is passed to Net::OpenSSH.  Net::OpenSSH defaults the username
                 to the current user.
   ssh_options : A hash which is passed to Net::OpenSSH.

=head2 $shell->run(@cmd)

Execute the command locally via IPC::Open3::Simple.  Calls the callback in
realtime for each line of output emitted from the command.

=head2 $ssh->run(@cmd)

Execute the command on a remote host via Net::OpenSSH.  Calls the callback in
realtime for each line of output emitted from the command.

=cut

sub shell {
    my ($class, %args) = @_;
    require Shell::Carapace::Shell;
    return Shell::Carapace::Shell->new(%args);
}

sub ssh {
    my ($class, %args) = @_;
    require Shell::Carapace::SSH;
    return Shell::Carapace::SSH->new(%args);
}

=head1 ABOUT THE NAME

Carapace: n. A protective, shell-like covering likened to that of a turtle or crustacean

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Eric Johnson E<lt>eric.git@iijo.orgE<gt>

=cut

1;
